core/src/main/scala/stryker4s/mutants/Mutator.scala

```diff
@@ -7,32 +7,85 @@ import fs2.io.file.Path
 import stryker4s.config.Config
 import stryker4s.extension.StreamExtensions._
 import stryker4s.log.Logger
-import stryker4s.model.{MutatedFile, MutationsInSource, SourceTransformations}
+import stryker4s.model._
 import stryker4s.mutants.applymutants.{MatchBuilder, StatementTransformer}
 import stryker4s.mutants.findmutants.MutantFinder
 
-import scala.meta.Tree
+import scala.meta._
 
 class Mutator(mutantFinder: MutantFinder, transformer: StatementTransformer, matchBuilder: MatchBuilder)(implicit
     config: Config,
     log: Logger
 ) {
 
-  def mutate(files: Stream[IO, Path]): IO[Seq[MutatedFile]] = {
+  def mutate(files: Stream[IO, Path], compileErrors: Seq[CompileError] = Seq.empty): IO[Seq[MutatedFile]] = {
     files
       .parEvalMapUnordered(config.concurrency)(p => findMutants(p).tupleLeft(p))
       .map { case (file, mutationsInSource) =>
-        val transformed = transformStatements(mutationsInSource)
-        val builtTree = buildMatches(transformed)
-
-        MutatedFile(file, builtTree, mutationsInSource.mutants, mutationsInSource.excluded)
+        val errorsInThisFile = compileErrors.filter(err => file.toString.endsWith(err.path))
+        mutateFile(file, mutationsInSource, errorsInThisFile)
       }
       .filterNot(mutatedFile => mutatedFile.mutants.isEmpty && mutatedFile.excludedMutants == 0)
       .compile
       .toVector
       .flatTap(logMutationResult)
   }
 
+  private def mutateFile(
+      file: Path,
+      mutationsInSource: MutationsInSource,
+      compileErrors: Seq[CompileError]
+  ): MutatedFile = {
+    val transformed = transformStatements(mutationsInSource)
+    val builtTree = buildMatches(transformed)
+
+    //If there are any compiler errors (i.e. we're currently retrying the mutation)
+    //Then we take the original tree built that didn't compile
+    //And then we search inside it to translate the compile errors to mutants
+    //Finally we rebuild it from scratch without those mutants
+    //This is not very performant, but you only pay the cost if there actually is a compiler error
+    val mutatedFile = MutatedFile(file, builtTree, mutationsInSource.mutants, Seq.empty, mutationsInSource.excluded)
+    if (compileErrors.isEmpty) {
+      mutatedFile
+    } else {
+      val nonCompilingIds = errorsToIds(compileErrors, mutatedFile.mutatedSource, mutationsInSource.mutants)
```

Comment:
Could you put this logic in a separate `RollbackHandler` or something similar? I think that'd keep this class a bit less cluttered as well as making it easier to test

Reply:
Done in 161a8d9